cpp/include/cudf/detail/utilities/hash_functions.cuh

```diff
@@ -66,7 +65,12 @@ void CUDA_DEVICE_CALLABLE md5_hash_step(md5_intermediate_data* hash_state,
         break;
     }
 
-    F = F + A + hash_constants[j] + buffer_ints[g];
+    uint32_t buffer_element_as_int;
+    thrust::copy_n(thrust::seq,
```

Comment:
I don't suggest using `thrust::copy` here. Use the compiler built-in `memcpy`, otherwise it may not get optimized away.

Reply:
@jrhemstad 
memcpy copies one byte at a time and uses loop. 
```
        ld.global.u8    %rs1, [%rd10];
        add.s64         %rd11, %rd3, %rd12;
        st.global.u8    [%rd11], %rs1;
```
`thrust::copy_n` copies the type size at a time. 
For example, for int it uses unrolled loop of copying 32-bit at a time. 
```
   ld.global.u32   %r3, [%rd4+8];
   st.global.u32   [%rd3+8], %r3;
   ld.global.u32   %r4, [%rd4+12];
```

Reply:
No actual copies should be performed. The compiler will optimize them away. This is a standard trick for type-punning without breaking aliasing rules: https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#how-do-we-type-pun-correctly

Reply:
Example: https://godbolt.org/z/xTrKnj

Reply:
It's optimizing away if these are registers! 
(for global memory, it doesn't, as expected).

Reply:
Yes, the compiler needs to see the declarations in order to elide the copies. The temporaries don't actually take up any registers.

Reply:
Changing this to `memcpy(&buffer_element_as_int, hash_state->buffer + g * 4, 4);` the compilation being killed
```
[22/76] Building CUDA object CMakeFiles/cudf.dir/src/hash/hashing.cu.o
FAILED: /usr/local/cuda-10.1/bin/nvcc  -Dcudf_EXPORTS -Igoogletest/install/include -I_deps/cub-src -I_deps/thrust-src -I_deps/jitify-src -I_deps/libcudacxx-src/include -I/usr/local/cuda-10.1/targets/x86_64-linux/include -Iinclude -Iinclude/jit -I../include -I../src -I/home/ryanlee/miniconda2/envs/cudf_dev_test/include -Xcompiler -Wno-parentheses -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_70,code=compute_70 --expt-extended-lambda --expt-relaxed-constexpr -Werror=cross-execution-space-call -Xcompiler -Wall,-Werror,-Wno-error=deprecated-declarations -Xcompiler -Wno-deprecated-declarations -DBOOST_NO_CXX14_CONSTEXPR -DHT_DEFAULT_ALLOCATOR -O3 -DNDEBUG -Xcompiler=-fPIC   -DJITIFY_USE_CACHE -DCUDF_VERSION=0.15.0 -std=c++14 -x cu -c ../src/hash/hashing.cu -o CMakeFiles/cudf.dir/src/hash/hashing.cu.o && /usr/local/cuda-10.1/bin/nvcc  -Dcudf_EXPORTS -Igoogletest/install/include -I_deps/cub-src -I_deps/thrust-src -I_deps/jitify-src -I_deps/libcudacxx-src/include -I/usr/local/cuda-10.1/targets/x86_64-linux/include -Iinclude -Iinclude/jit -I../include -I../src -I/home/ryanlee/miniconda2/envs/cudf_dev_test/include -Xcompiler -Wno-parentheses -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_70,code=compute_70 --expt-extended-lambda --expt-relaxed-constexpr -Werror=cross-execution-space-call -Xcompiler -Wall,-Werror,-Wno-error=deprecated-declarations -Xcompiler -Wno-deprecated-declarations -DBOOST_NO_CXX14_CONSTEXPR -DHT_DEFAULT_ALLOCATOR -O3 -DNDEBUG -Xcompiler=-fPIC   -DJITIFY_USE_CACHE -DCUDF_VERSION=0.15.0 -std=c++14 -x cu -M ../src/hash/hashing.cu -MT CMakeFiles/cudf.dir/src/hash/hashing.cu.o -o CMakeFiles/cudf.dir/src/hash/hashing.cu.o.d
Killed
```

Maybe i'm missing something here, but that's why I had gone back to `thrust::copy_n` despite your original suggestion.

Reply:
:/ That will require further investigation.

Reply:
After removing the `md5_element_hasher` class, the replacing the copy_n worked fine.